feature-find-meal/src/main/java/com/kinandcarta/feature/find/meal/viewmodel/DisplayMealsViewModel.kt

```diff
@@ -0,0 +1,122 @@
+package com.kinandcarta.feature.find.meal.viewmodel
+
+import android.location.Location
+import android.util.Log
+import androidx.annotation.RequiresPermission
+import androidx.hilt.lifecycle.ViewModelInject
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.google.android.gms.location.FusedLocationProviderClient
+import com.google.android.gms.location.LocationCallback
+import com.google.android.gms.location.LocationRequest
+import com.google.android.gms.location.LocationResult
+import com.google.android.gms.maps.model.LatLng
+import com.kcc.kmmhackathon.shared.MealsSDK
+import com.kcc.kmmhackathon.shared.entity.Meal
+import com.kcc.kmmhackathon.shared.utility.DistanceUnit
+import kotlinx.coroutines.launch
+
+typealias Meals = List<Meal>
+
+class DisplayMealsViewModel @ViewModelInject constructor(
+    private val fusedLocationProviderClient: FusedLocationProviderClient
+) : ViewModel() {
+
+    sealed class State {
+        object LoadingMeals : State()
+        data class LoadedMeals(val meals: Meals, val distanceUnit: DistanceUnit) : State()
+        data class ReservedMeal(val code: String, val position: Int) : State()
+        data class LocationUpdate(val userLatLng: LatLng) : State()
+        data class MealUnavailable(val code: String) : State()
+        data class Failed(val failure: Failure) : State()
+    }
+
+    sealed class Failure(cause: Throwable) : Throwable(cause) {
+        class LoadingMealsFailed(cause: Throwable) : Failure(cause)
+        class ReserveAMealFailed(cause: Throwable) : Failure(cause)
+    }
+
+    val state: LiveData<State> get() = _state
+
+    private val _state: MutableLiveData<State> = MutableLiveData(State.LoadingMeals)
+    private var lastLocation: Location? = null
+    private val sdk = MealsSDK()
+    private val locationCallback = object : LocationCallback() {
+        override fun onLocationResult(locationResult: LocationResult?) {
+            lastLocation = locationResult?.lastLocation
+            updateMeals("onLocationResult")
+        }
+    }
+
+    override fun onCleared() {
+        fusedLocationProviderClient.removeLocationUpdates(locationCallback)
+    }
+
+    @RequiresPermission("android.permission.ACCESS_FINE_LOCATION")
+    fun startUpdatingLocation() {
+        fusedLocationProviderClient.lastLocation.addOnSuccessListener {
+            if (it != lastLocation) {
+                lastLocation = it
+                updateUserLocation()
+            }
+            updateMeals("start updating location")
+        }
+        fusedLocationProviderClient.requestLocationUpdates(
+            createLocationRequest(),
+            locationCallback,
+            null
+        )
+    }
+
+    private fun createLocationRequest(): LocationRequest {
+        val locationRequest = LocationRequest()
+        locationRequest.interval = 100000
```

Comment:
When I made one of the ViewModel PRs Roger suggested I be more clear about the intervals by adding a local variable:

```
    private fun createLocationRequest(): LocationRequest {
        val oneMinuteFortySecondsInMs: Long = 100000
        val fiveSecondsInMs: Long = 5000
        val locationRequest = LocationRequest()
        locationRequest.interval = oneMinuteFortySecondsInMs
        locationRequest.fastestInterval = fiveSecondsInMs
        locationRequest.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
        return locationRequest
    }

```

Reply:
Oh yep! Good shout 😊